tests/system_tests/test_artifacts/test_wandb_artifacts_api.py

```diff
@@ -241,6 +241,52 @@ def test_log_artifact_with_invalid_tags(tmp_path, user, wandb_init, api, invalid
             run.log_artifact(artifact, tags=invalid_tags)
 
 
+def test_retrieve_artifacts_by_tags(user, wandb_init, server_supports_artifact_tags):
+    project = "test"
+    artifact_name = "test-artifact"
+    artifact_type = "test-type"
+
+    with wandb_init(entity=user, project=project) as run:
+        for i in range(10):
+            artifact = wandb.Artifact(name=artifact_name, type=artifact_type)
+            with artifact.new_file(f"{i}.txt", "w") as f:
+                f.write("testing")
+            run.log_artifact(artifact)
+
+    artifact_name = f"{user}/{project}/{artifact_name}"
+
+    for logged_artifact in Api().artifacts(type_name=artifact_type, name=artifact_name):
+        version = int(logged_artifact.version.strip("v"))
+        if version in [2, 3, 5, 7]:
+            logged_artifact.tags.append("prime")
+        if version % 3 == 0:
+            logged_artifact.tags.append("fizz")
+        if version % 5 == 0:
+            logged_artifact.tags.append("buzz")
+        logged_artifact.save()
+
+    # Retrieve all artifacts with a given tag.
+    artifacts = Api().artifacts(
+        type_name=artifact_type, name=artifact_name, tags=["prime"]
```

Comment:
@moredatarequired - qq - I thought we had discussed a more generic filtering arg like filter where we only support tags as part of this project. This is similar to the [runs API](https://docs.wandb.ai/ref/python/public-api/api#runs) where 
<img width="640" alt="Screenshot 2024-09-19 at 9 00 46 PM" src="https://github.com/user-attachments/assets/123c96c8-8ba4-4aa4-8b81-cbbd82c976e5">

In the future, [here](https://www.notion.so/wandbai/Registry-Product-Roadmap-20104da87dbf496496a90c01fa8112b7?pvs=4#316ba628d0e44b648edde26bf84b508b) are some other fields we might want to filer on. Again, not supporting all of them now but it's going to be messy to deprecate the tags param if in a few months we want to add a filter by alias, for example.

Reply:
Follow up, summarizing from offline discussion:

The `api.artifacts` call requires a fully qualified artifact, so it doesn't make sense to search for name, prefix, or type (those are fixed) and if you have the alias you have direct access to that version with `api.artifact`. So the other attributes we've discussed filtering on aren't relevant here.

We _can_ expose the `filters` parameter, and might in the future, but filtering on tags is not as intuitive as passing in tags directly (it requires something like `filters = {"$and": [{"tags.name": tag1}, {"tags.name": tag2}]}`), so it's probably better from a UI perspective to do it this way.

Reply:
Adding color -- for tags, since it's a `list[str]`, it gets more complicated and unintuitive.  If we're sticking with raw, dict-like mongodb filters, the operators you probably want to use are
- `$all` to require **all** tags match 
  - e.g. `{"tags": {"$all": ["tag1", "tag2"]}}`
- `$elemMatch`+`$in` to require **any** tags match 
  - e.g. `{"tags": {"$elemMatch": {"$in": ["tag1", "tag2"]}}}`

`$all` is supported [in the backend](https://github.com/wandb/core/blob/d53a4fab5b232578ca888d35214122fce8c08f25/services/gorilla/pkg/filtering/filter.go#L571-L580), not sure about `$elemMatch` (or if this is handled through different operators)